test/Npgsql.Tests/Types/RangeTests.cs

```diff
@@ -1,123 +1,73 @@
+            .SetName("EmptyRange"),
+        new TestCaseData(new NpgsqlRange<decimal>(1, true, 10, true), "[1,10]", "numrange", NpgsqlDbType.NumericRange)
+            .SetName("Inclusive"),
+        new TestCaseData(new NpgsqlRange<decimal>(1, false, 10, false), "(1,10)", "numrange", NpgsqlDbType.NumericRange)
+            .SetName("Exclusive"),
+        new TestCaseData(new NpgsqlRange<decimal>(1, true, 10, false), "[1,10)", "numrange", NpgsqlDbType.NumericRange)
+            .SetName("InclusiveExclusive"),
+        new TestCaseData(new NpgsqlRange<decimal>(1, false, 10, true), "(1,10]", "numrange", NpgsqlDbType.NumericRange)
+            .SetName("ExclusiveInclusive"),
+        new TestCaseData(new NpgsqlRange<decimal>(1, false, true, 10, false, false), "(,10)", "numrange", NpgsqlDbType.NumericRange)
+            .SetName("InfiniteLowerBound"),
+        new TestCaseData(new NpgsqlRange<decimal>(1, true, false, 10, false, true), "[1,)", "numrange", NpgsqlDbType.NumericRange)
+            .SetName("InfiniteUpperBound")
+    };
+
+    [Test, TestCaseSource(nameof(RangeTestCases))]
+    public Task Range<T>(T range, string sqlLiteral, string pgTypeName, NpgsqlDbType? npgsqlDbType)
+        => AssertType(range, sqlLiteral, pgTypeName, npgsqlDbType);
 
     [Test]
-    public async Task Range()
-    {
-        using var conn = await OpenConnectionAsync();
-        using var cmd = new NpgsqlCommand("SELECT @p1, @p2, @p3, @p4", conn);
-        var p1 = new NpgsqlParameter("p1", NpgsqlDbType.Range | NpgsqlDbType.Integer) { Value = NpgsqlRange<int>.Empty };
-        var p2 = new NpgsqlParameter { ParameterName = "p2", Value = new NpgsqlRange<int>(1, 10) };
-        var p3 = new NpgsqlParameter { ParameterName = "p3", Value = new NpgsqlRange<int>(1, false, 10, false) };
-        var p4 = new NpgsqlParameter { ParameterName = "p4", Value = new NpgsqlRange<int>(0, false, true, 10, false, false) };
-        Assert.That(p2.NpgsqlDbType, Is.EqualTo(NpgsqlDbType.Range | NpgsqlDbType.Integer));
-        cmd.Parameters.Add(p1);
-        cmd.Parameters.Add(p2);
-        cmd.Parameters.Add(p3);
-        cmd.Parameters.Add(p4);
-        using var reader = await cmd.ExecuteReaderAsync();
-        reader.Read();
-
-        Assert.That(reader[0].ToString(), Is.EqualTo("empty"));
-        Assert.That(reader[1].ToString(), Is.EqualTo("[1,11)"));
-        Assert.That(reader[2].ToString(), Is.EqualTo("[2,10)"));
-        Assert.That(reader[3].ToString(), Is.EqualTo("(,10)"));
-    }
-
+    public Task DateRange_with_DateTime()
+        => AssertType(new NpgsqlRange<DateTime>(
+                new DateTime(2020, 1, 1, 0, 0, 0), true,
+                new DateTime(2020, 1, 3, 0, 0, 0), false),
+            "[2020-01-01,2020-01-03)", "daterange", NpgsqlDbType.DateRange,
+            isDefaultForWriting: false);
+
+#if NET6_0_OR_GREATER
     [Test]
-    [NonParallelizable]
-    public async Task Range_with_long_subtype()
```

Comment:
User-defined range support seems broken in the PR. The following test works in main but fails in this PR with `No configured writer was found for parameter of type NpgsqlTypes.NpgsqlRange`1[System.Double] and DataTypeName 'temp_type1'.`:

```c#
[Test]  
public async Task User_defined_range()  
{  
    await using var connection = await OpenConnectionAsync();  
    var rangeTypeName = await GetTempTypeName(connection);  
    await connection.ExecuteNonQueryAsync($"CREATE TYPE {rangeTypeName} AS RANGE (SUBTYPE = double precision)");  
  
    await using var dataSource = CreateDataSource();  
    await AssertType(  
        dataSource,        new NpgsqlRange<double>(3.5d, 4.5d),  
        "[3.5,4.5]",  
        rangeTypeName,        npgsqlDbType: null, // TODO: NpgsqlDbType should work, but we don't support inferring the data type from it (that gets tested)  
        isNpgsqlDbTypeInferredFromClrType: false);  
}
```

Reply:
> I'm a bit surprised we don't have tests for it

We did have one, I accidentally removed it while doing the work here :grimacing: 

Agreed for NpgsqlDbType, I forget we had already discussed this.

Reply:
Addressed in https://github.com/npgsql/npgsql/pull/5123/commits/13cb873ac075b4144c05a2bd1bb3925bf12d8921